### PR TITLE
chore: Update to TLDraw 2.0

### PR DESCRIPTION
**Pull request by richburdon**
- [x] Upgrade to 2.1.4
- [x] Auto-migrate data on-the-fly

- ISSUE: Need to go read-only if detect other peers have migrated the document and we are behind.

https://github.com/dxos/dxos/assets/3523355/b059fe32-d9f9-4d27-9c9c-d4d053b04004

### CODE REVIEW

**Comment by jessmartin** [comment]
@richburdon I started on this as well, just note that we need to not land this until we handle migration, as described in #6099

File: packages/apps/plugins/plugin-sketch/src/hooks/adapter.ts [hunk 1]
```diff
@@ -259,5 +280,58 @@ const decode = (value: any): any => {
...
+          store.put([record]);
+        } catch (err) {
+          log.catch(err, { record });
+        }
+      }
+    }
+  }
+};
+
+// TODO(burdon): Need unit test and check that records are duplicated.
```
Comment (wittjosiah) [inline packages/apps/plugins/plugin-sketch/src/hooks/adapter.ts:324]: I don't know enough about the tldraw impl to assess this too much, but I'd say based on our other experiences with migrating data that grabbing some real sketches from existing prod spaces and testing this on them manually would be a good sanity check before shipping this.'

**Review by jessmartin — APPROVED** [review]

File: packages/apps/plugins/plugin-sketch/project.json [hunk 2]
```diff
@@ -19,9 +19,12 @@
         ]
       }
     },
-    "lint": {}
+    "lint": {},
+    "test": {},
```
[the remark this answers is not in the document]
Comment (richburdon) [reply packages/apps/plugins/plugin-sketch/project.json:23]: i couldn't get vitest to work.

File: packages/apps/plugins/plugin-sketch/src/components/Sketch.stories.tsx [hunk 3]
```diff
@@ -13,13 +13,15 @@ import { Button, Toolbar } from '@dxos/react-ui';
 
 import SketchComponent from './Sketch';
 
+const createSketch = () => create(SketchType, { data: create(Expando, {}) });
```
[the remark this answers is not in the document]
Comment (richburdon) [reply packages/apps/plugins/plugin-sketch/src/components/Sketch.stories.tsx:16]: yes

File: packages/apps/plugins/plugin-sketch/src/hooks/adapter.ts (hunk 1 above)
Comment (richburdon) [reply packages/apps/plugins/plugin-sketch/src/hooks/adapter.ts:325]: grabbing = copy and paste right? that works.
i just noticed though that copying from tldraw.com (which i assume is 2.0) doesn't work.

**Comment by jessmartin** [comment]
Is the grid and snap-to-grid a setting that can be toggled?

**Review by dmaretskyi — APPROVED** [review]

**Comment by jessmartin** [comment]
@richburdon Testing Notes:

OP = Composer peer with old TLDraw (main)
NP = Composer peer with new TLDraw (running `burdon/tldraw-2` branch) 

## Test 1: Old sketch file gets automatically updated

- Created a new space with OP
- Created a sketch file with OP and put some edits in it
- Invited NP into the same space
- Opened the sketch file with NP
- File is read-only for NP and missing a few elements (arrows and lines - see image)

Screenshot 2024-05-21 at 2 50 47 PM

My expectation is that opening the sketch file with NP would automatically migrate the file and OP would be in read-only mode.

## Test 2: New Sketch files in existing spaces are created on the new version

- Created a new space with OP
- Invited NP into the same space
- Created a new sketch file with NP
- File is read-only for NP and editable by OP

My expectation is that the sketch file would be editable by NP and read-only for OP.

## Test 3: New Sketch files in new spaces are created on the new version

- Created a new space with NP
- Created a new sketch file with NP
- File is editable by NP
- Invited OP into the same space
- File remains editable by NP and crashes OP ("Content failed to render.")
 
image

## Test 4: Old Sketch files get automatically updated to the new version (single reader/writer)

- Created a new space with OP
- Created a sketch file with OP and put some edits in it
- Updated the OP to the new code (by switching branch) and restarted Composer
- Opened the sketch file with peer running code on branch (technically, a NP, but same client)
- File is read-only, uneditable, and has some missing content (arrows, lines)

My expectation is that the file should have been automatically migrated and should now be editable.

## Other Problems

- [ ] controls and grid disappear jarringly whenever the mouse leaves the immediate canvas area

    https://github.com/dxos/dxos/assets/27258/8a7d87e3-0df9-4503-91e8-d382341d462e

- [ ] "Move to page" is available in the context menu when selecting some options, but there's no way to navigate between pages

    image

File: packages/apps/plugins/plugin-sketch/src/components/Sketch.tsx [hunk 4]
```diff
@@ -30,83 +25,116 @@ export type SketchComponentProps = {
...
+      <Tldraw
+        // Setting the key forces re-rendering when the content changes.
+        key={sketch.id}
+        store={adapter.store}
+        hideUi={!active}
+        // TODO(burdon): Customize assets: https://tldraw.dev/docs/assets
+        maxAssetSize={1024 * 1024}
+        components={{
+          DebugPanel: null,
+          Grid: CustomGrid,
```
Comment (jessmartin) [inline packages/apps/plugins/plugin-sketch/src/components/Sketch.tsx:101]: What if we don't like this new grid and prefer the TLDraw-standard grid? Can this be a setting?

File: packages/apps/plugins/plugin-sketch/src/components/Sketch.tsx [hunk 5]
```diff
@@ -30,83 +25,116 @@ export type SketchComponentProps = {
...
-        styles.background,
-        className,
-      )}
+      className={mx('is-full bs-full', className)}
+      onPointerEnter={() => {
+        setActive(!readonly && !adapter.readonly);
+      }}
+      onPointerLeave={() => {
+        setActive(false);
+      }}
```
Comment (jessmartin) [inline packages/apps/plugins/plugin-sketch/src/components/Sketch.tsx:89]: This looks like maybe it was added to make Sketch work better in Stacks/Collections, but it's really awkward when Sketch is presented in the `main` surface.

**Pull request closed without merging**